core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1300,20 +1346,13 @@ private void setBranchSnapshotInternal(Snapshot snapshot, String branch) {
       this.lastUpdatedMillis = isAddedSnapshot(snapshot.snapshotId()) ?
           snapshot.timestampMillis() : System.currentTimeMillis();
 
-      if (SnapshotRef.MAIN_BRANCH.equals(branch)) {
```

Comment:
The lines that set `lastUpdatedMillis` also need to move to ensure that `lastUpdatedMillis` is set correctly.

Reply:
Okay, thinking about this, I think that it is slightly different for the other method. I think it should be this:

```java
if (isAddedSnapshot(snapshot.snapshotId()) {
  this.lastUpdatedMillis = snapshot.timestampMillis();
}

if (SnapshotRef.MAIN_BRANCH.equals(branch)) {
  this.currentSnapshotId = replacementSnapshotId;
  if (lastUpdatedMillis == null) {
    this.lastUpdatedMillis = System.currentTimeMillis();
  }
  snapshotLog.add(new SnapshotLogEntry(lastUpdatedMillis, replacementSnapshotId));
}
```

That ensures that if a branch is updated to a newly added snapshot, `lastUpdatedMillis` is set to match the new snapshot, just like we do now with main. In addition, if `main` is set to an existing snapshot, that will use the same timestamp for the commit and for snapshot history. And if either of those isn't the case---main isn't changing and there isn't a new snapshot---then the update time is set when the metadata is built.

Reply:
Good point, I completely missed about the lastUpdated time. We can remove the existing logic in setBranchSnapshotInternal and make it contained within setRef. We would anyways want that because in case somebody updates main to a new snapshot we should also update the last updated time.

Ok there are 3 cases here:

1.) New snapshot ( on any branch) we set the the lastUpdatedMillis to be the "current time"
2.) Main snapshot gets updated. We should set the last updated only if it hasn't previously been set (due to new snapshots)
3.) If set ref isn't called. As you mentioned here then we delegate to the build time lastUpdatedMillis